File: frontend/src/api/projects.ts
import { apiJson } from './client';

export type EntityType = 'gene' | 'variant' | 'disease' | 'mixed';

export interface ProjectSnapshot {
  id: string;
  project_id: string;
  name?: string | null;
  state: Record<string, unknown>;
  created_at: string;
}

export interface SavedProject {
  id: string;
  title: string;
  description?: string | null;
  entity_type: EntityType;
  query: string;
  tags: string[];
  is_archived: boolean;
  created_at: string;
  updated_at: string;
  latest_snapshot?: ProjectSnapshot | null;
}

export interface ProjectCreatePayload {
  title: string;
  description?: string;
  entity_type: EntityType;
  query: string;
  tags: string[];
  state: Record<string, unknown>;
}

export const projectsApi = {
  list: () => apiJson<SavedProject[]>('/api/projects'),
  create: (payload: ProjectCreatePayload) =>
    apiJson<SavedProject>('/api/projects', { method: 'POST', body: JSON.stringify(payload) }),
  get: (id: string) => apiJson<SavedProject>(`/api/projects/${id}`),
  archive: (id: string) => apiJson<{ message: string }>(`/api/projects/${id}`, { method: 'DELETE' }),
};

File: backend/app/schemas/projects.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


EntityType = Literal["gene", "variant", "disease", "mixed"]


class ProjectSnapshotCreate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    state: dict[str, Any]


class ProjectSnapshotResponse(BaseModel):
    id: str
    project_id: str
    name: str | None = None
    state: dict[str, Any]
    created_at: datetime

    model_config = {"from_attributes": True}


class ProjectCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    entity_type: EntityType
    query: str = Field(min_length=1, max_length=255)
    tags: list[str] = Field(default_factory=list)
    state: dict[str, Any]


class ProjectUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    entity_type: EntityType | None = None
    query: str | None = Field(default=None, min_length=1, max_length=255)
    tags: list[str] | None = None


class ProjectResponse(BaseModel):
    id: str
    title: str
    description: str | None = None
    entity_type: EntityType
    query: str
    tags: list[str]
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    latest_snapshot: ProjectSnapshotResponse | None = None

    model_config = {"from_attributes": True}
